exp/operator/controllers/shared_test.go

```diff
@@ -0,0 +1,253 @@
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+				&operatorv1.InfrastructureProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "infra-2",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+			},
+			expectedCondition: clusterv1.Condition{
+				Type:   operatorv1.PreflightCheckCondition,
+				Status: corev1.ConditionTrue,
+			},
+		},
+		{
+			name: "two different infra providers exist in different namespaces, preflight check passed",
+			providers: []client.Object{
+				&operatorv1.InfrastructureProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "infra-3",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+				&operatorv1.InfrastructureProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "infra-4",
+						Namespace: namespaceName2,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+			},
+			expectedCondition: clusterv1.Condition{
+				Type:   operatorv1.PreflightCheckCondition,
+				Status: corev1.ConditionTrue,
+			},
+		},
+		{
+			name: "two similar infra provider exist in different namespaces, preflight check failed",
+			providers: []client.Object{
+				&operatorv1.InfrastructureProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "infra-3",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+				&operatorv1.InfrastructureProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "infra-3",
+						Namespace: namespaceName2,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "InfrastructureProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+			},
+			expectedCondition: clusterv1.Condition{
+				Type:     operatorv1.PreflightCheckCondition,
+				Reason:   operatorv1.MoreThanOneExistsReason,
+				Severity: clusterv1.ConditionSeverityWarning,
+				Message:  operatorv1.MoreThanOneExistsMessage,
+				Status:   corev1.ConditionFalse,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			gs := NewWithT(t)
+
+			for _, c := range tc.providers {
+				// Controller runtime removes gvk after Create(), so we need to save it
+				// and set back after
+				gvk := c.GetObjectKind().GroupVersionKind()
+				gs.Expect(testEnv.Create(ctx, c)).To(Succeed())
+				c.GetObjectKind().SetGroupVersionKind(gvk)
+			}
+
+			defer func() {
+				testEnv.Cleanup(ctx, tc.providers...)
```

Comment:
What did we decide on cleanup? Does it work per test case or do we need to delete and wait?

Reply:
Each test creates its own resources, they should all be cleaned up in the end.